wormchain/x/wormhole/keeper/msg_server_wasmd_test.go

```diff
@@ -221,4 +237,169 @@ func TestWasmdInstantiateContract(t *testing.T) {
+	_ = privateKeys
+	k.SetConfig(ctx, types.Config{
+		GovernanceEmitter:     vaa.GovernanceEmitter[:],
+		GovernanceChain:       uint32(vaa.GovernanceChain),
+		ChainId:               uint32(vaa.ChainIDWormchain),
+		GuardianSetExpiration: 86400,
+	})
+	signer_bz := [20]byte{}
+	signer := sdk.AccAddress(signer_bz[:])
+
+	set := createNewGuardianSet(k, ctx, guardians)
+
+	context := sdk.WrapSDKContext(ctx)
+	msgServer := keeper.NewMsgServerImpl(*k)
+
+	// First we need to (1) upload some codes and (2) instantiate.
+	// (1) upload
+	payload := createWasmStoreCodePayload(keepertest.EXAMPLE_WASM_CONTRACT_GZIP)
+	code_ids := []uint64{}
+	for i := 0; i < 5; i++ {
+		v := generateVaa(set.Index, privateKeys, vaa.ChainID(vaa.GovernanceChain), payload)
+		vBz, err := v.Marshal()
+		assert.NoError(t, err)
+		res, err := msgServer.StoreCode(context, &types.MsgStoreCode{
+			Signer:       signer.String(),
+			WASMByteCode: keepertest.EXAMPLE_WASM_CONTRACT_GZIP,
+			Vaa:          vBz,
+		})
+		assert.NoError(t, err)
+
+		code_ids = append(code_ids, res.CodeID)
+	}
+
+	// (2) instantiate
+	payload = createWasmInstantiatePayload(code_ids[0], "btc", "{}")
+	v := generateVaa(set.Index, privateKeys, vaa.ChainID(vaa.GovernanceChain), payload)
+	vBz, _ := v.Marshal()
+	instantiate, err := msgServer.InstantiateContract(context, &types.MsgInstantiateContract{
+		Signer: signer.String(),
+		CodeID: code_ids[0],
+		Label:  "btc",
+		Msg:    []byte("{}"),
+		Vaa:    vBz,
+	})
+	require.NoError(t, err)
+
+	// Now we can test migrating
+
+	// Confirm migrate works
+	for _, code_id := range code_ids {
+		payload = createWasmMigratePayload(code_id, instantiate.Address, "{}")
+		v = generateVaa(set.Index, privateKeys, vaa.ChainID(vaa.GovernanceChain), payload)
+		vBz, _ = v.Marshal()
+		_, err = msgServer.MigrateContract(context, &types.MsgMigrateContract{
+			Signer:   signer.String(),
+			CodeID:   code_id,
+			Contract: instantiate.Address,
+			Msg:      []byte("{}"),
+			Vaa:      vBz,
+		})
+		require.NoError(t, err)
+	}
+
+	// Test failure using the wrong codeid
+	payload = createWasmMigratePayload(code_ids[0], instantiate.Address, "{}")
+	v = generateVaa(set.Index, privateKeys, vaa.ChainID(vaa.GovernanceChain), payload)
+	vBz, _ = v.Marshal()
+	_, err = msgServer.MigrateContract(context, &types.MsgMigrateContract{
+		Signer: signer.String(),
+		// Switch codeid
+		CodeID:   code_ids[1],
+		Contract: instantiate.Address,
+		Msg:      []byte("{}"),
+		Vaa:      vBz,
+	})
+	assert.ErrorIs(t, err, types.ErrInvalidHash)
+
+	// Test failure using the wrong contract
+	payload = createWasmMigratePayload(code_ids[0], instantiate.Address, "{}")
+	v = generateVaa(set.Index, privateKeys, vaa.ChainID(vaa.GovernanceChain), payload)
+	vBz, _ = v.Marshal()
+	_, err = msgServer.MigrateContract(context, &types.MsgMigrateContract{
+		Signer: signer.String(),
+		CodeID: code_ids[0],
+		// modify address
+		Contract: instantiate.Address + "a",
```

Comment:
here the address itself will probably fail validation. This does not necessarily test the address check.